api/src/main/java/ai/djl/translate/LMSearch.java

```diff
@@ -0,0 +1,617 @@
+                        searchState.pastAttentionMask.concat(
+                                manager.ones(
+                                        new Shape(inputIds.getShape().get(0), 1), DataType.INT64),
+                                1);
+
+                // memory management
+                NDScope.unregister(
+                        searchState.nextInputIds,
+                        searchState.pastAttentionMask,
+                        searchState.pastOutputIds);
+                NDScope.unregister(searchState.pastKeyValues);
+            }
+
+            // Termination Criteria
+            // TODO: <EOS>, delete the sentence and add it to result.
+            if (searchState.pastOutputIds.getShape().get(1) + 1 >= config.maxSeqLength) {
+                break;
+            }
+        }
+        return searchState.pastOutputIds.concat(searchState.nextInputIds, 1);
+    }
+
+    // https://huggingface.co/blog/how-to-generate
+    public NDArray beamSearch(NDArray inputIds) {
+        NDArray attentionMask = prepareAttentionMaskOffset(inputIds, config);
+        NDManager manager = inputIds.getManager();
+        long numBeam = config.beam;
+        long numBatch = inputIds.getShape().get(0);
+        BeamSearchState searchState = new BeamSearchState();
+
+        long numHeads = 0;
+        long kvDim = 0;
+        while (true) {
+            if (searchState.pastAttentionMask == null) {
+                // Initial beams
+                NDList modelInput = prepareInput(inputIds, attentionMask, 0, 1);
+                CausalLMOutput modelOutput = lmBlock.forward(modelInput, null, manager);
+
+                // [batch, probDim]
+                NDArray allProbs = modelOutput.logits.get(":, -1, :").softmax(1);
+
+                // [batch, beam]
+                NDList topK = allProbs.topK((int) numBeam, -1, true, false);
+                NDArray outputIds = topK.get(1).expandDims(2);
+                NDArray lastProbs = topK.get(0).normalize(1, 1);
+                assert outputIds.getShape().getShape().length == 3 : "Wrong shape";
+                assert lastProbs.getShape().getShape().length == 2 : "Wrong Shape";
+
+                // [batch, beam, seq + 1]
+                attentionMask =
+                        attentionMask
+                                .concat(manager.ones(new Shape(numBatch, 1), DataType.INT64), -1)
+                                .expandDims(1)
+                                .repeat(1, numBeam);
+
+                // [batch, beam, heads, seq_past, kvFeature]
+                Function<NDArray, NDArray> fn = ndarray -> ndarray.expandDims(1).repeat(1, numBeam);
+                NDList pastKeyValues =
+                        new NDList(
+                                modelOutput.pastKeyValuesList.stream()
+                                        .map(fn)
+                                        .collect(Collectors.toList()));
+                // [batch, beam, seq_past]
```

Comment:
These shape comments are a great addition! It makes it much easier to track the NDArray operations